internal/envstest/server.go

```diff
@@ -0,0 +1,208 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package envstest
+
+import (
+	"context"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/google/exposure-notifications-server/pkg/keys"
+	"github.com/google/exposure-notifications-server/pkg/server"
+	"github.com/google/exposure-notifications-verification-server/internal/project"
+	"github.com/google/exposure-notifications-verification-server/internal/routes"
+	"github.com/google/exposure-notifications-verification-server/pkg/cache"
+	"github.com/google/exposure-notifications-verification-server/pkg/config"
+	"github.com/google/exposure-notifications-verification-server/pkg/database"
+	"github.com/google/exposure-notifications-verification-server/pkg/ratelimit"
+	"github.com/sethvargo/go-envconfig"
+	"github.com/sethvargo/go-limiter"
+	"github.com/sethvargo/go-limiter/memorystore"
+)
+
+// TestServerResponse is used as the reply to creating a test UI server.
+type TestServerResponse struct {
+	Config      *config.ServerConfig
+	Database    *database.Database
+	Cacher      cache.Cacher
+	KeyManager  keys.KeyManager
+	RateLimiter limiter.Store
+	Server      *server.Server
+}
+
+// NewServer creates a new test UI server instance. When this function returns,
+// a full UI server will be running locally on a random port. Cleanup is handled
+// automatically.
+func NewServer(tb testing.TB) *TestServerResponse {
```

Comment:
It returns the server and everything needed to run it. We have to return things like the database because tests might need to insert or assert records.

Reply:
I'm open to other names